lib/logstash/filters/aggregate.rb

```diff
@@ -0,0 +1,228 @@
+	DEFAULT_TIMEOUT = 1800
+
+	# This is the state of the filter.
+	# For each entry, key is "task_id" and value is a map freely updatable by 'code' config
+	@@aggregate_maps = {}
+
+	# Mutex used to synchronize access to 'aggregate_maps'
+	@@mutex = Mutex.new
+
+	# Aggregate instance which will evict all zombie Aggregate elements (older than timeout)
+	@@eviction_instance = nil
+
+	# last time where eviction was launched
+	@@last_eviction_timestamp = nil
+
+	# Initialize plugin
+	public
+	def register
+		# process lambda expression to call in each filter call
+		eval("@codeblock = lambda { |event, map| #{@code} }", binding, "(aggregate filter code)")
+
+		# define eviction_instance
+		@@mutex.synchronize do
+			if (@timeout > 0 && (@@eviction_instance.nil? || @timeout < @@eviction_instance.timeout))
+				@@eviction_instance = self
+				@logger.info("Aggregate, timeout: #{@timeout} seconds")
+			end
+		end
+	end
+
+	
+	# This method is invoked each time an event matches the filter
+	public
+	def filter(event)
+		# return nothing unless there's an actual filter event
+		return unless filter?(event)
+
+		# define task id
+		task_id = event.sprintf(@task_id)
+		return if task_id.nil? || task_id.empty? || task_id == @task_id
+
+		@@mutex.synchronize do
+			# retrieve the current aggregate map
+			aggregate_maps_element = @@aggregate_maps[task_id]
+			if (aggregate_maps_element.nil?)
+				return if @map_action == "update"
+				aggregate_maps_element = LogStash::Filters::Aggregate::Element.new(Time.now);
+				@@aggregate_maps[task_id] = aggregate_maps_element
+			else
+				return if @map_action == "create"
+			end
+			map = aggregate_maps_element.map
+
+			# execute the code to read/update map and event
+			@codeblock.call(event, map)
+			
+			# delete the map if task is ended
+			@@aggregate_maps.delete(task_id) if @end_of_task
+		end
+
+		filter_matched(event)
+	end
+
+	# Necessary to indicate logstash to periodically call 'flush' method
+	def periodic_flush
+		true
+	end
+  
+	# This method is invoked by LogStash every 5 seconds.
+	def flush(options = {})
+		# Protection against no timeout defined by logstash conf : define a default eviction instance with timeout = DEFAULT_TIMEOUT seconds
+		if (@@eviction_instance.nil?)
+			@@eviction_instance = self
+			@timeout = DEFAULT_TIMEOUT
+		end
+		
+		# Launch eviction only every interval of (@timeout / 2) seconds
+		if (@@eviction_instance == self && (@@last_eviction_timestamp.nil? || Time.now > @@last_eviction_timestamp + @timeout / 2))
+			remove_expired_elements()
+			@@last_eviction_timestamp = Time.now
+		end
+		
+		return nil
+	end
+
+	
+	# Remove the expired Aggregate elements from "aggregate_maps" if they are older than timeout
+	def remove_expired_elements()
+		min_timestamp = Time.now - @timeout
+		@@mutex.synchronize do
+			@@aggregate_maps.delete_if { |key, element| element.creation_timestamp < min_timestamp }
+		end
+	end
+
+	# Getter/Setter methods used for the tests
+	def self.aggregate_maps
```

Comment:
This should be removed,  if you want to access them from a test class you can use something like this http://apidock.com/ruby/Module/class_variable_get